lowfi_simulation/src/test/native/cpp/MotorEncoderSimulatorTest.cpp

```diff
@@ -0,0 +1,126 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2018 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "Encoder.h"
+#include "MotorEncoderSimulator.h"
+#include "MotorModel/LinearMotorModelSimulation.h"
+#include "Talon.h"
+#include "WpiSimulators/WpiEncoderSimulator.h"
+#include "WpiSimulators/WpiMotorSimulator.h"
+#include "gtest/gtest.h"
+
+TEST(MotorEncoderSimulatorTest, TestWithoutDistancePerPulseFullSpeed) {
+  frc::Talon talon{3};
+  frc::Encoder encoder{3, 1};
+  std::shared_ptr<frc::sim::lowfi::WpiMotorSimulator> motorSim(
```

Comment:
Why std::shared_ptr? For user-facing code at least, we've been tending toward stack-based allocation and reference parameters in constructors. It's more idiomatic C++, and it's also less to type and harder for users to get wrong.

For example:
```
frc::sim::lowfi::WpiMotorSimulator motorSim(3);
frc::sim::lowfi::LinearMotorModelSimulation motorModelSim(6000);
motorSim.SetMotorModelSimulation(motorModelSim);
frc::sim::lowfi::WpiEncoderSimulator encoderSim(0);
frc::sim::lowfi::MotorEncoderSimulator connector(motorSim, encoderSim);
```
Our C++ APIs would probably go like this if you only set the simulation model once:
```
frc::sim::lowfi::LinearMotorModelSimulation motorModelSim(6000);
frc::sim::lowfi::WpiMotorSimulator motorSim(3, motorModelSim);
frc::sim::lowfi::WpiEncoderSimulator encoderSim(0);
frc::sim::lowfi::MotorEncoderSimulator connector(motorSim, encoderSim);
```
The names with "simulation" and "simulator" on the end feel kind of long. Doesn't "LinearMotorModel" get the point across? In general, I'm concerned that the difference between simulator and simulation isn't entirely obvious out of context. The code shows how a simulation object goes into a simulator, but couldn't you also reasonably expect a simulator to produce simulations rather than consume them? I think by "simulation", you really mean "model".

Reply:
It depends on if you want/need to store the reference for use elsewhere.  SnobotSim keeps a collection of all of the "standard" simulator wrappers  So in order to do that, the connector would either need to hold on to a reference and the caller would need to make sure it does not go out of scope.

Reply:
At least for all the WPILib classes so far, teams have never needed to pass them into a foreign scope that outlives it. If the object being created and being passed out has Robot class scope like a subsystem, it should get created as a member of the Robot class anyway. I haven't encountered a legitimate case where lifetimes of dependent hardware objects weren't strictly nested.